File: backend/agents/recipe_generator_agent.py
"""Recipe generator agent using mock mode or OpenAI mode."""

import os

from services.openai_service import OpenAIConfigError, OpenAIJSONError, OpenAIService


class RecipeGeneratorAgent:
    """Generates recipes from pantry + user preferences."""

    def __init__(self) -> None:
        self.openai_service = OpenAIService()

    def _generate_mock_recipe(
        self,
        pantry_items: list[dict],
        meal_type: str,
        preference: str,
        use_only_pantry: bool,
        user_message: str,
    ) -> dict:
        pantry_names = {item["name"] for item in pantry_items}
        base_ingredients = []
        for item in pantry_items[:4]:
            base_ingredients.append(
                {
                    "name": item["name"],
                    "amount": min(float(item["amount"]), 200),
                    "unit": item["unit"],
                }
            )

        if not base_ingredients:
            base_ingredients = [{"name": "egg", "amount": 2, "unit": "unit"}]

        missing_ingredients = []
        if not use_only_pantry and "salt" not in pantry_names:
            missing_ingredients.append({"name": "salt", "amount": 2, "unit": "g"})

        return {
            "title": f"Mock {preference.title()} {meal_type.title()} Bowl",
            "ingredients": base_ingredients,
            "instructions": [
                "Prepare ingredients from the list.",
                "Cook and combine ingredients in a pan or pot.",
                "Serve warm.",
                f"User request considered: {user_message.strip()}" if user_message.strip() else "",
            ],
            "servings": 1,
            "tags": [preference, meal_type, "mock"],
            "missing_ingredients": missing_ingredients,
        }

    def _build_openai_prompt(
        self,
        pantry_items: list[dict],
        meal_type: str,
        preference: str,
        use_only_pantry: bool,
        user_message: str,
    ) -> str:
        return f"""
You are a recipe generator. Return JSON only (no markdown, no explanations).

Required JSON schema:
{{
  "title": "Recipe title",
  "ingredients": [{{"name":"ingredient","amount":100,"unit":"g"}}],
  "instructions": ["Step 1", "Step 2"],
  "servings": 1,
  "tags": ["high protein", "lunch"],
  "missing_ingredients": []
}}

Rules:
- Allowed units only: g, ml, unit.
- Forbidden units: cup, cups, tbsp, tsp, pinch, handful, to taste.
- Every ingredient must include name, amount, unit.
- If use_only_pantry is true, use only pantry ingredients.
- If use_only_pantry is false, you may suggest missing ingredients and list them in missing_ingredients.

Inputs:
- pantry_items: {pantry_items}
- meal_type: {meal_type}
- preference: {preference}
- use_only_pantry: {use_only_pantry}
- user_message: {user_message}
"""

    def generate_recipe(
        self,
        pantry_items: list[dict],
        meal_type: str,
        preference: str,
        use_only_pantry: bool,
        user_message: str,
    ) -> dict:
        use_mock_openai = os.getenv("USE_MOCK_OPENAI", "false").lower() == "true"
        if use_mock_openai:
            recipe = self._generate_mock_recipe(
                pantry_items, meal_type, preference, use_only_pantry, user_message
            )
            recipe["instructions"] = [s for s in recipe["instructions"] if s]
            return recipe

        prompt = self._build_openai_prompt(
            pantry_items, meal_type, preference, use_only_pantry, user_message
        )
        try:
            return self.openai_service.generate_recipe_json(prompt)
        except (OpenAIConfigError, OpenAIJSONError):
            raise
        except Exception as exc:
            raise RuntimeError(str(exc))

File: supabase/functions/api/agents/recipeGeneratorAgent.ts
import type { PantryItem } from "./pantryAgent.ts";
import { generateRecipeWithOpenAI, getOpenAIModel } from "../services/openaiService.ts";

export interface GenerateInput {
  pantryItems: PantryItem[];
  mealType: string;
  preference: string;
  useOnlyPantry: boolean;
  message: string;
  servings: number;
}

export interface RecipeDraft {
  title: string;
  description: string;
  ingredients: Array<{ name: string; amount: number; unit: "g" | "ml" | "unit" }>;
  instructions: string[];
  servings: number;
  tags: string[];
  missing_ingredients: Array<{ name: string; amount: number; unit: "g" | "ml" | "unit" }>;
}

function mockRecipe(input: GenerateInput): RecipeDraft {
  const servings = Math.max(1, Math.floor(input.servings || 1));
  const preferred = input.preference || "quick meal";
  const mealType = input.mealType || "meal";
  const pantryByName = new Map(input.pantryItems.map((i) => [i.name, i]));
  const pick = (name: string, fallbackAmount: number, unit: "g" | "ml" | "unit") => {
    const p = pantryByName.get(name);
    if (!p) return { name, amount: fallbackAmount, unit };
    return { name, amount: Math.min(Number(p.amount), fallbackAmount), unit: p.unit as "g" | "ml" | "unit" };
  };
  const base = [pick("egg", 2 * servings, "unit"), pick("rice", 90 * servings, "g"), pick("milk", 120 * servings, "ml")];
  const missing = [] as Array<{ name: string; amount: number; unit: "g" | "ml" | "unit" }>;
  if (!input.useOnlyPantry && !pantryByName.has("chicken")) missing.push({ name: "chicken", amount: 120 * servings, unit: "g" });
  return {
    title: `${preferred} ${mealType} bowl`.replace(/\b\w/g, (c) => c.toUpperCase()),
    description: `Mock recipe generated for ${mealType} with ${preferred}. ${input.message || ""}`.trim(),
    ingredients: base.filter((i) => i.amount > 0),
    instructions: ["Prepare ingredients and measure quantities.", "Cook main ingredients until done.", "Combine, season lightly, and serve warm."],
    servings,
    tags: [preferred, mealType, "mock"],
    missing_ingredients: input.useOnlyPantry ? [] : missing,
  };
}

export async function generateRecipe(input: GenerateInput): Promise<RecipeDraft> {
  const mock = String(Deno.env.get("USE_MOCK_OPENAI") ?? "true").toLowerCase() === "true";
  if (mock) return mockRecipe(input);

  const systemPrompt = `Return strict JSON only. No markdown. No extra text.
JSON schema:
{"title":"...","description":"...","ingredients":[{"name":"...","amount":100,"unit":"g"}],"instructions":["..."],"servings":1,"tags":["..."],"missing_ingredients":[]}
Rules:
- Exactly requested servings.
- Quantities are total recipe quantities.
- Realistic per-serving quantities.
- Pantry amounts are limits only.
- If use_only_pantry=true: use only pantry items and do not exceed pantry amounts.
- If use_only_pantry=false: missing ingredients allowed and must appear in missing_ingredients.
- Allowed units only: g, ml, unit.
- Forbidden units: cups, tbsp, tsp, pinch, handful, to taste.
- Avoid absurd quantities and do not use whole pantry unless appropriate.`;

  const userPrompt = JSON.stringify({
    request: {
      meal_type: input.mealType,
      preference: input.preference,
      use_only_pantry: input.useOnlyPantry,
      message: input.message,
      servings: input.servings,
    },
    pantry_items: input.pantryItems,
  });

  const json = await generateRecipeWithOpenAI(systemPrompt, userPrompt);
  return json as unknown as RecipeDraft;
}

export { getOpenAIModel };
